piccolo_api/media/azure.py

```diff
@@ -0,0 +1,432 @@
+                self.connection_string
+            )
+        else:
+            account_url = (
+                f"https://{self.storage_account_name}.blob.core.windows.net"
+            )
+
+            # DefaultAzureCredential is not part of the BlobServiceClient types
+            default_credential = t.cast(str, DefaultAzureCredential())
+
+            blob_service_client = BlobServiceClient(
+                account_url, credential=default_credential
+            )
+
+        container_client = blob_service_client.get_container_client(
+            self.container_name
+        )
+
+        return container_client, blob_service_client
+
+    async def store_file(
+        self, file_name: str, file: t.IO, user: t.Optional[BaseUser] = None
+    ) -> str:
+        loop = asyncio.get_running_loop()
+
+        blocking_function = functools.partial(
+            self.store_file_sync, file_name=file_name, file=file, user=user
+        )
+
+        file_key = await loop.run_in_executor(self.executor, blocking_function)
+
+        return file_key
+
+    def _prepend_folder_name(self, file_key: str) -> str:
+        folder_name = self.folder_name
+        if folder_name:
+            return str(pathlib.Path(folder_name, file_key))
+        else:
+            return file_key
+
+    def store_file_sync(
+        self, file_name: str, file: t.IO, user: t.Optional[BaseUser] = None
+    ) -> str:
+        """
+        A sync wrapper around :meth:`store_file`.
+        """
+        file_key = self.generate_file_key(file_name=file_name, user=user)
+        extension = file_key.rsplit(".", 1)[-1]
+        client, _ = self.get_client()
+        upload_metadata: t.Dict[str, t.Any] = self.upload_metadata.copy()
+
+        if extension in CONTENT_TYPE:
+            upload_metadata["ContentType"] = CONTENT_TYPE[extension]
+
+        blob_client = client.get_blob_client(
+            self._prepend_folder_name(file_key)
+        )
+
+        blob_client.upload_blob(data=file, metadata=upload_metadata)
+
+        return file_key
+
+    async def generate_file_url(
+        self, file_key: str, root_url: str, user: t.Optional[BaseUser] = None
+    ) -> str:
+        """
+        This retrieves an absolute URL for the file.
+        """
+        loop = asyncio.get_running_loop()
+
+        blocking_function: t.Callable = functools.partial(
+            self.generate_file_url_sync,
+            file_key=file_key,
+            root_url=root_url,
+            user=user,
+        )
+
+        return await loop.run_in_executor(self.executor, blocking_function)
+
+    def generate_file_url_sync(
+        self, file_key: str, root_url: str, user: t.Optional[BaseUser] = None
+    ) -> str:
+        """
+        A sync wrapper around :meth:`generate_file_url`.
+        """
+        blob_name = self._prepend_folder_name(file_key)
+
+        if not self.sign_urls:
+            return (
+                f"https://{self.storage_account_name}.blob.core.windows.net/"
```

Comment:
Is it worth having `blob.core.windows.net` as a class variable, or someone can override this somehow if needed? Just in case there are other services which are API compatible with Azure storage.

Reply:
Yeah good shout, its rare but it _would_ change if dealing with GovClouds